### PR TITLE
Rholang: Integrate Tuplespace with Evaluator 

### PR DESCRIPTION
**Pull request by KentShikama**
Relative to #447 and #445 and block by persistent flag in storage.

### CODE REVIEW

File: rholang/src/test/scala/coop/rchain/rholang/interpreter/ReduceTest.scala [hunk 1]
```diff
@@ -102,45 +122,49 @@ class ReduceSpec extends FlatSpec with Matchers {
...
-      _      <- receiveTask
-      _      <- sendTask
-      tsHash <- interpreter.tupleSpace.spaceMVar.take
-      _      <- interpreter.tupleSpace.spaceMVar.put(HashMap())
+      tupleSpace  <- Task.pure(InMemoryStore.create[Channel, List[Channel], List[Quote], Par])
+      interpreter <- Task.pure(Reduce.makeInterpreter(tupleSpace))
+      _           <- interpreter.eval(receive)(Env())
+      _           <- interpreter.eval(send)(Env())
+      tsHash      <- Task.pure(tupleSpace.toHashMap)
     } yield tsHash
```
Comment (KentShikama) [inline rholang/src/test/scala/coop/rchain/rholang/interpreter/ReduceTest.scala]: @henrytill Any Scala expertise on how I can recreate the InMemoryStore each time without having to create a new test?

File: storage/src/test/scala/coop/rchain/storage/test/InMemoryStore.scala [hunk 2]
```diff
@@ -34,6 +34,18 @@ class InMemoryStore[C, P, A, K] private (
...
+
+  def toHashMap: mutable.HashMap[List[C], ((List[A], Boolean), (List[(List[P], K)], Boolean))] = {
+    val hashMap =
+      mutable.HashMap.empty[List[C], ((List[A], Boolean), (List[(List[P], K)], Boolean))]
+    keys.map(key =>
+      withTxn(createTxnRead()) { txn =>
+        hashMap += ((key, ((getAs(txn, key), false), (getPsK(txn, key), false))))
+    })
+    hashMap
+  }
```
Comment (KentShikama) [inline storage/src/test/scala/coop/rchain/storage/test/InMemoryStore.scala]: So I'm thinking that keys() can be exposed but toHashMap will just be internal.

**Comment by henrytill** [comment]
> and block by persistent flag in storage

That would be #449.

File: rholang/src/main/scala/coop/rchain/rholang/interpreter/Reduce.scala [hunk 3]
```diff
@@ -178,9 +92,17 @@ object Reduce {
     def consume(binds: Seq[(Seq[Channel], Quote)], body: Par, persistent: Boolean)(
         implicit env: Env[Par]): Task[Option[Cont[Par, Par]]] =
       binds match {
-        case Nil         => Task raiseError new Error("Error: empty binds")
-        case bind +: Nil => tupleSpace.simpleConsume(bind._2, bind._1, body, persistent)
-        case _           => Task raiseError new Error("Error: joins are not currently implemented.")
+        case Nil => Task raiseError new Error("Error: empty binds")
+        case bind +: Nil =>
+          internalConsume(tupleSpace, List(Channel(bind._2)), List(bind._1.toList), body) match {
```
Comment (guardbotmk3) [inline rholang/src/main/scala/coop/rchain/rholang/interpreter/Reduce.scala]: I don't see an environment here.

File: rholang/src/main/scala/coop/rchain/rholang/interpreter/Reduce.scala [hunk 4]
```diff
@@ -40,112 +48,9 @@ trait Reduce[M[_], C, A, P, B] {
...
-          }
-        }
-        _ <- spaceMVar.put(space)
-      } yield result
-  }
 
-  // implicit def DebruijnInterpreter: Reduce[Task, Quote, Par, Channel, Par] =
-  class DebruijnInterpreter extends Reduce[Task, Quote, Par, Channel, Par] {
-    val tupleSpace = new SimpleTuplespace()
+  class DebruijnInterpreter(implicit tupleSpace: IStore[Channel, List[Channel], List[Quote], Par])
```
[the remark this answers is not in the document]
Comment (KentShikama) [reply rholang/src/main/scala/coop/rchain/rholang/interpreter/Reduce.scala]: Done

File: rholang/src/main/scala/coop/rchain/rholang/interpreter/Reduce.scala [hunk 5]
```diff
@@ -29,7 +37,7 @@ import scala.collection.mutable.HashMap
   */
 trait Reduce[M[_], C, A, P, B] {
   type Cont[Data, Body] = (Body, Env[Data])
-  def produce(chan: C, data: Seq[A], persistent: Boolean)(
+  def produce(chan: C, data: List[A], persistent: Boolean)(
```
[the remark this answers is not in the document]
Comment (KentShikama) [reply rholang/src/main/scala/coop/rchain/rholang/interpreter/Reduce.scala]: Done

File: rholang/src/main/scala/coop/rchain/rholang/interpreter/Reduce.scala [hunk 6]
```diff
@@ -160,9 +65,18 @@ object Reduce {
...
-        implicit env: Env[Par]): Task[Option[Cont[Par, Par]]] =
-      tupleSpace.simpleProduce(chan, data, persistent)
+    def produce(chan: Quote, data: List[Par], persistent: Boolean)(
+        implicit env: Env[Par]): Task[Option[Cont[Par, Par]]] = {
+      // TODO: Handle the environment in the store
+      val substData: List[Quote] = data.map(p => Quote(substitute(p)(env)))
+      internalProduce(tupleSpace, Channel(chan), substData) match {
+        case Some((body, dataList)) =>
+          val newEnv: Env[Par] =
+            Env.makeEnv(dataList.flatMap(identity).map({ case Quote(p) => p }): _*)
```
[the remark this answers is not in the document]
Comment (KentShikama) [reply rholang/src/main/scala/coop/rchain/rholang/interpreter/Reduce.scala]: Will resolve with bandaid fix.

File: rholang/src/main/scala/coop/rchain/rholang/interpreter/Reduce.scala (hunk 3 above)
Comment (KentShikama) [reply rholang/src/main/scala/coop/rchain/rholang/interpreter/Reduce.scala]: Will resolve with bandaid fix.

File: rholang/src/main/scala/coop/rchain/rholang/interpreter/Reduce.scala [hunk 7]
```diff
@@ -178,9 +92,17 @@ object Reduce {
...
+        case bind +: Nil =>
+          internalConsume(tupleSpace, List(Channel(bind._2)), List(bind._1.toList), body) match {
+            case Some((body, dataList)) =>
+              val newEnv: Env[Par] =
+                Env.makeEnv(dataList.flatMap(identity).map({ case Quote(p) => p }): _*)
+              Task.pure(Some((body, newEnv)))
+            case None => Task.pure(None)
+          }
+        // TODO: join should be almost free with internalConsume but test cases still need to be written
+        case _ => Task raiseError new Error("Error: joins are not currently implemented.")
```
[the remark this answers is not in the document]
Comment (KentShikama) [reply rholang/src/main/scala/coop/rchain/rholang/interpreter/Reduce.scala]: Will do

File: rholang/src/test/scala/coop/rchain/rholang/interpreter/ReduceTest.scala [hunk 8]
```diff
@@ -13,14 +13,18 @@ import scala.collection.immutable.BitSet
...
-
 import Reduce.Cont
+import coop.rchain.storage.Serialize
+import coop.rchain.storage.test.InMemoryStore
 
 class ReduceSpec extends FlatSpec with Matchers {
+  implicit val serializer = Serialize.mkProtobufInstance(Channel)
+
   "evalExpr" should "handle simple addition" in {
+    implicit val tupleSpace = InMemoryStore.create[Channel, List[Channel], List[Quote], Par]
```
Comment (guardbotmk3) [inline rholang/src/test/scala/coop/rchain/rholang/interpreter/ReduceTest.scala]: Why is this still implicit?

**Comment by KentShikama** [comment]
Ready for another review.

File: rholang/src/test/scala/coop/rchain/rholang/interpreter/ReduceTest.scala (hunk 8 above)
Comment (KentShikama) [reply rholang/src/test/scala/coop/rchain/rholang/interpreter/ReduceTest.scala]: Done

File: storage/src/main/scala/coop/rchain/storage/internal.scala [hunk 9]
```diff
@@ -1,12 +1,10 @@
...
 
-private[storage] final case class Datum[A](a: A, persist: Boolean)
+final case class Datum[A](a: A, persist: Boolean)
 
 private[storage] final case class DataCandidate[C, A](channel: C, datum: Datum[A], datumIndex: Int)
 
-private[storage] final case class WaitingContinuation[P, K](patterns: List[P],
-                                                            continuation: K,
-                                                            persist: Boolean)
+final case class WaitingContinuation[P, K](patterns: List[P], continuation: K, persist: Boolean)
```
Comment (KentShikama) [inline storage/src/main/scala/coop/rchain/storage/internal.scala]: Note that I've discussed with @henrytill and we are going to expose these two case classes until we come up with a better solution.

**Pull request merged**